tests/pkg/cspc/cspcspecbuilder/specbuilder.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2020 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cspcspecbuilder
+
+import (
+	cstor "github.com/openebs/api/pkg/apis/cstor/v1"
+	"github.com/openebs/cstor-operators/tests/pkg/cache/cspccache"
+	"github.com/openebs/cstor-operators/tests/pkg/infra"
+	"k8s.io/klog"
+)
+
+type CSPCSpecBuilder struct {
+	CSPCCache    *cspccache.CSPCResourceCache
+	Infra        *infra.Infrastructure
+	CSPC         *cstor.CStorPoolCluster
+	CSPCSpecData *CSPCSpecData
+}
+
+type CSPCSpecData struct {
+	UsedNodes   map[string]bool
+	UnUsedNodes map[string]bool
+	UsedDisks   map[string]bool
+	UnUsedDisk  map[string]bool
+}
+
+func NewCSPCSpecData() *CSPCSpecData {
+	return &CSPCSpecData{
+		UsedNodes:   map[string]bool{},
+		UnUsedNodes: map[string]bool{},
+		UsedDisks:   map[string]bool{},
+		UnUsedDisk:  map[string]bool{},
+	}
+}
+
+func (cd *CSPCSpecData) AddNodeToUsedSet(nodeName string) {
+	cd.UsedNodes[nodeName] = true
+	delete(cd.UnUsedNodes, nodeName)
+}
+
+func (cd *CSPCSpecData) AddDiskToUsedSet(diskName string) {
+	cd.UsedDisks[diskName] = true
+	delete(cd.UnUsedDisk, diskName)
+}
+
+func (cd *CSPCSpecData) AddNodeToUnusedSet(nodeName string) {
+	cd.UnUsedNodes[nodeName] = true
+	delete(cd.UsedNodes, nodeName)
+}
+
+func (cd *CSPCSpecData) AddDiskToUnusedSet(diskName string) {
+	cd.UnUsedDisk[diskName] = true
+	delete(cd.UsedDisks, diskName)
+}
+
+func NewCSPCSpecBuilder(cspcCache *cspccache.CSPCResourceCache, infra *infra.Infrastructure) *CSPCSpecBuilder {
+	// Initialize CSPCSpecData
+	cspcSpecData := NewCSPCSpecData()
+	for _, nodeName := range cspcCache.NodeList {
+		cspcSpecData.UnUsedNodes[nodeName] = true
+		for _, bd := range cspcCache.NodeDisk[nodeName] {
+			cspcSpecData.UnUsedDisk[bd] = true
```

Comment:
do we need to add check here? If we are calling this NewCSPCSpecBuilder multiple times then this comment is valid
```go
if _, ok := cspcSpecData.UsedDisks; !ok { // add into UnusedDisk }
```